Blue green deployments 2 (#4403)
Kong/docs.konghq.com: src/gateway/upgrade.md

The upgrade to 3.0.x is a **major** upgrade.
The lowest version that Kong 3.0.x supports migrating from is 2.1.x.

{:.important}

> > **Important**: Blue-green migration for versions below 2.8.2 to 3.0.x is not supported.
> The upcoming 2.8.2 release will include blue-green migration support. If you want
> to perform migrations with no downtime, please wait for the upcoming 2.8.2 patch release,
> upgrade to 2.8.2, [then migrate to 3.0.x](#migrate-db).

Reviewer: ```suggestion
> **Important**: Blue-green migration for traditional mode for versions below 2.8.2 to 3.0.x is not supported.
The upcoming 2.8.2 release will include blue-green migration support. If you want
to perform migrations for traditional mode with no downtime, please wait for the upcoming 2.8.2 patch release,
upgrade to 2.8.2, [then migrate to 3.0.x](#migrate-db).
```

Amazon Linux 1 and Debian 8 (Jessie) containers and packages are deprecated and are no longer produced for new versions of {{site.base_gateway}}.

#### Blue-green deployments

> **Traditional mode**: Blue-green upgrades from versions of 2.8.1.x and below 3.0.0.0 are not currently supported for traditional mode.

Reviewer 2: ```suggestion
**Traditional mode**: Blue-green upgrades from versions 2.8.1.x and below to 3.0.0.0 are not currently supported, which means we can not migrate with no downtime for traditional mode.
```

Amazon Linux 1 and Debian 8 (Jessie) containers and packages are deprecated and are no longer produced for new versions of {{site.base_gateway}}.

Blue-green deployments:
* **{{site.ee_product_name}}**: Blue-green upgrades from versions 2.1.x.x-2.7.x.x to 3.0.0.0 are not currently supported.

>   * This is a known issue planned to be fixed in the next 2.8.x.x release. When that version is released 2.x users should upgrade to that version before beginning a blue-green upgrade to 3.0.0.0.

Reviewer: ```suggestion
  * This is a known issue planned to be fixed in the next 2.8.x.x release. When that version is released, 2.x users should upgrade to that version before beginning a blue-green upgrade to 3.0.0.0.
```

Amazon Linux 1 and Debian 8 (Jessie) containers and packages are deprecated and are no longer produced for new versions of {{site.base_gateway}}.

Blue-green deployments:

> * Traditional mode: Blue-green upgrades from versions of 2.8.1(2.8.1.x for Enterprise) and below to 3.0.0(3.0.0.0 for Enterprise) are not currently supported for traditional mode.

Reviewer: ```suggestion
* Traditional mode: Blue-green upgrades from versions of 2.8.1.x and below to 3.0.0.0 are not currently supported for traditional mode.
```

  * This is a known issue planned to be fixed in the next 2.8 release. When that version is released, 2.x users should upgrade to that version before beginning a blue-green upgrade to 3.0.

> * Hybrid mode: migration can be done by following steps:

Reviewer: ```suggestion
* Hybrid mode: you can migrate Enterprise hybrid mode deployments with the following method:
```

>   1. Upgrade control plane (override installation and reload, or stop and start);
>   2. Start new version of data planes;
>   3. gracefully shut down old version of data planes.
> 
>   This solution works in 2 ways for different setup:
>   1. For 2.8 dataplanes of Enterprise, new control plane will downgrade the configuration when syncing up;
>   2. For other situation(open-source, or version below 2.8), the control plane won't sync up any new configuration
>   to old data planes.
> 

Reviewer: ```suggestion
  1. Upgrade the control plane: Override the previous installation with version 3.0.0.0 and reload, or stop and start {{site.kong_gateway}}.
  2. Install {{site.base_gateway}} 3.0.0.0 on new data planes.
  3. Gracefully shut down old versions of data planes.

  Depending on your setup, this solution will have the following effects:
  1. For 2.8 data planes of {{site.ee_product_name}}, the new control plane will downgrade the configuration when syncing up.
  2. For other situations (open-source, or versions below 2.8), the control plane won't sync up any new configuration to old data planes.

```

> Note that either way, the blue-green deployment should be a temporary status, and we should fully transfer to the
> new version when it's ready. We recommend not to change configurations when applying blue-green migration, as it may
> leads to issues:
> * For traditional mode, old instances may create old format of entities to configuration, and new instances may create entities that old instances don't understand, and thus causes unexpected behavior;
> * For hybrid mode, old dataplanes may not response to the change of the configuration.

Reviewer: ```suggestion
Note that either way, the blue-green deployment should be a temporary state, and you should fully transfer to the
new version when it's ready. We recommend not changing configurations when using blue-green migration, as it may
leads to issues:
* For traditional mode, old instances may create conflicts with old entity formats in configuration, and new instances may create entities that old instances don't understand, causing unexpected behavior.
* For hybrid mode, old data planes may not respond to the change in the configuration.
```

Amazon Linux 1 and Debian 8 (Jessie) containers and packages are deprecated and are no longer produced for new versions of {{site.base_gateway}}.

Blue-green deployments:
* Traditional mode: Blue-green upgrades from versions of 2.8.1(2.8.1.x for Enterprise) and below to 3.0.0(3.0.0.0 for Enterprise) are not currently supported for traditional mode.
  * This is a known issue planned to be fixed in the next 2.8 release. When that version is released, 2.x users should upgrade to that version before beginning a blue-green upgrade to 3.0.

> * Hybrid mode: migration can be done by following steps:

Reviewer 3: Does it make sense to describe this process for EE customers in this brief form?  it might be better to place this description in an internal document that is used by the sustaining team to help customers who actually need to do a blue-green upgrade of EE before we've made the 2.8.x helper release.

Reviewer: > it might be better to place this description in an internal document that is used by the sustaining team to help customers who actually need to do a blue-green upgrade of EE before we've made the 2.8.x helper release.

We don't have/maintain internal documents for the sustaining team. If you're aware of any location of these docs, you're welcome to use + copy any of this content to add to them 😄

Reviewer 2: > Does it make sense to describe this process for EE customers in this brief form? it might be better to place this description in an internal document that is used by the sustaining team to help customers who actually need to do a blue-green upgrade of EE before we've made the 2.8.x helper release.

Yes those steps are better put into `upgrade.md`.

You should not configure with admin API during migration, as it may lead to unexpected behavior and
break your configuration.

**Version prerequisites for migrating to version 3.0.x**

If you are migrating from 2.7.x or lower version, you should first migrate to 2.8.1.
Here's how you can migrate to 2.8.1: [Upgrade from 1.0.x - 2.2.x to 2.8.x](#upgrade-from-10x---22x-to-28x)

Once you migrated to 2.8.x, you can follow the instructions in the section
below to migrate to 3.0.x.

### Upgrade from 2.8.x (x>=2) to 3.0.x for Traditional mode

1. Clone your database.
2. Download 3.0.x, and configure it to point to the cloned datastore
   as your old (2.8.2 or beyond) cluster. Run `kong migrations up` and `kong migrations finish`.
3. Start 3.0.x cluster.
4. Now both the old (2.8.x) and new (3.0.x)
   clusters can now run simultaneously. Start provisioning 3.0.x nodes.
3. Gradually divert traffic away from your old nodes, and into
   your 3.0.x cluster. Monitor your traffic to make sure everything
   is going smoothly.
4. When your traffic is fully migrated to the 3.0.x cluster,
   decommission your old nodes.

> ### Upgrade to 3.0.x for Hybrid mode

Reviewer: ```suggestion
### Upgrade to 3.0.x for hybrid mode
```

Data planes are capable to serve traffic during the process of migration. 

1. Download 3.0.x.
2. Decommission your old control plane.

> 3. Configure new control plane to point to the same datastore

Reviewer: ```suggestion
3. Configure the new control plane to point to the same data store
```

   as your old control plane. Run `kong migrations up` and `kong migrations finish`.

> 4. Start new control plane. It is expected that old data planes may complain

Reviewer: ```suggestion
4. Start the new control plane. It is expected that old data planes may complain
```

{:.important}
> **Important**:

> TODO confirm this, and if false, remove this line. If true we need to change all the below documentation to reference EE-only instructions...

Author: 🚧 🚧 🚧 
I want to get the answer to this before merging this PR

Reviewer 3: I think this is mostly a product/political decision.  Blue/green upgrades on CE would require a compatibility 2.8.x release that enables 2.8.x DP nodes to work after the `up` migration part has run.  We're going to make such a release for EE, but I don't think that the same is planned for CE.

Author: Talked with Rob - I think there was miscommunication. CE does support no-downtime (blue/green) upgrades, but does not support 3.0 CP to 2.8 DP compatibility. Users could shut down their 2.8 CP but leave the 2.8 DPs serving traffic while bringing up 3.0 CP and DP nodes, then migrate traffic to the 3.0 cluster to do a no downtime deploy.

{:.important}
> **Important**:

> TODO confirm this, and if false, remove this line. If true we need to change all the below documentation to reference EE-only instructions...

Reviewer: ```suggestion
```
